### DISCUSSION TITLE
How can I figure out the required compute version for GPUs?

### DISCUSSION THREAD

**Discussion by ukd1** [ultralytics/yolov5#9676] [Q&A] [answered]
I'm writing a project using yolov5, and it runs fast on my Mac - but slow on my server, which has no GPU. I'm looking at some second hand Tesla GPUs on eBay, but I'm not entirely sure what version is going to be required. For ~$100 you can get a Tesla M4, and for $185 a Tesla P4.

Looking [here](https://developer.nvidia.com/cuda-gpus) and or [here](https://en.wikipedia.org/wiki/CUDA) lists the M4 as compute `5.2`...and tpe P4 as `6.1`. These, from what I understand are the compute version supported.

So, yolov5 [seems to use 2.4.2](https://github.com/ultralytics/yolov5/blob/c98128fe71a8676037a0605ab389c7473c743d07/requirements.txt#L33) and the tensor flow [lists](https://www.tensorflow.org/install/source#gpu) says that supports compute `8.0` and CUDA SDK `11.0`.

So, then looking at CUDA [release notes](https://docs.nvidia.com/cuda/archive/11.5.0/cuda-toolkit-release-notes/index.html), they say that SDK `11.5` – `11.7.1` support for compute capability `3.5` – `8.7` (Kepler (in part), Maxwell, Pascal, Volta, Turing, Ampere)...

AKA; the SDK would work on either of those GPUs...and that's the SDK required by the project's tensor flow...

However, I whilst it seems safe to presume that that version of tensorflow allows using GPUs in range of compute capability from `3.5 - 8.7`, I'm really unclear how to tell if a project like this has a specific requirements; e.g. it uses some features from compute `7`, and therefor that's the min...

So a couple of questions:
* what is the min compute support for a GPU I need?
* how can I tell this for a project using tensorflow?

**Comment by glenn-jocher** [comment, marked as answer]
@ukd1 good question. In general you'll get the most computer per $$ from the consumer GPUs (the Tesla models are enterprise GPUs that are much more expensive). In the aftermarket area the 1080Ti is a great deal, or if you have a slightly higher budget maybe a newer 2080Ti or 3080Ti are great. Most recent GPUs will run well with current versions of torch and tf.

Also you should always export to TensorRT at FP16 for inference on CUDA devices, you'll see about 5x speedup vs torch on GPU.

**Reply by ukd1** [reply]
Thanks @glenn-jocher; I was mostly looking at those older Tesla cards as lower-powered cards that were passively cooled for some servers I run. Fans can cause more maintainance. I'll def see if I can find passive 1080Tis though, hadn't thought of that.

Eitherway can you help with the specifics, as I'm still unclear:
* what is the min compute support for a GPU I need?
* how can I tell this for a project using tensorflow?

**Reply by glenn-jocher** [reply]
@ukd1 we don't track min compute as it's never really been an issue. You can train YOLOv5 in Colab with a K80 GPU for example which was launched in 2014 without any problems, so I'm assuming anything newer than that will work as well.

**Reply by ukd1** [reply]
@glenn-jocher roger that. Any idea if that translates to what I can run (without having to retrain) the existing models on?

**Reply by ukd1** [reply]
Basically, I'm trying to find a cheaper way to run faster than CPU that fits in 1u server; did consider a TPU, but figure a GPU might be more verstaile.

**Reply by glenn-jocher** [reply]
@ukd1 yeah that's a good question. OctoML has some cool comparisons of price vs performance for different clouds: https://try.octoml.ai/yolov5/

On the passively cooled side Apple's been doing amazing work with their M1/M2 chips. The latest M2 macbooks can run YOLOv5s in about 10 ms with --device MPS

For the cheapest/fastest U1 solution with modern cards T4's a very hard to beat, but like I said for much cheaper GPUs older 1080Ti's are extremely cost effective.

**Reply by ukd1** [reply]
@glenn-jocher thanks - I'm running it in dev on an M1 Max, and it's nicely fast - 70ms...but on CPU like 500ms+. Whilst I somehow have a spare M1, I kinda wanna run it on Kubernetes - which I run on Linux. I should look in to if I can get k8s to run on it, and expose the GPU.

T4s would def fit, and I looked, but are ~$1000 it seems. Even if they're faster (a lot faster!), I'm not sure I can justify the 5-10x increase at the mo. Def gonna checkout the consumer cards.

Any thoughts on TPU? The corals, even though jacked price-wise, are still around $100/ea on eBay for a PCIe version. Downside to me seems to be the inability to use for training, only inference?

**Reply by glenn-jocher** [reply]
You can use an Edge TPU with TFLite, but you'll get much better results with a Jetson nano with TensorRT at that price point.

**Reply by ukd1** [reply]
As it's not able to run a decent sized version of yolo? Nano, will check it out if I don't go the GPU route.

**Reply by glenn-jocher** [reply]
@ukd1 not sure, we don't have any edge TPUs ourselves, but you can export any model to edge TPU format easily, i.e.:
```
python export.py --weights yolov5x.pt --include edge_tpu
```
